qttypes/src/lib.rs

```diff
@@ -875,6 +883,336 @@ mod tests {
+            return self->contains(key);
+        })
+    }
+
+    /// Wrapper around [`clear()`][method] method.
+    ///
+    /// [method]: https://doc.qt.io/qt-5/qmap.html#clear
+    pub fn clear(&mut self) {
+        cpp!(unsafe [self as "QVariantMap*"] {
+            self->clear();
+        })
+    }
+
+    /// Wrapper around [`value(const QString &, const QVariant &)`][method] method.
+    ///
+    /// [method]: https://doc.qt.io/qt-5/qmap.html#value
+    pub fn value(&self, key: QString, default_value: QVariant) -> QVariant {
+        cpp!(unsafe [self as "QVariantMap*", key as "QString", default_value as "QVariant"] -> QVariant as "QVariant" {
+            return self->value(key, default_value);
+        })
+    }
+
+    /// Wrapper around [`key(const QVariant &, const QString &)`][method] method.
+    ///
+    /// [method]: https://doc.qt.io/qt-5/qmap.html#key
+    pub fn key(&self, value: QVariant, default_key: QString) -> QString {
+        cpp!(unsafe [self as "QVariantMap*", default_key as "QString", value as "QVariant"] -> QString as "QString" {
+            return self->key(value, default_key);
+        })
+    }
+}
+
+impl Index<QString> for QVariantMap {
+    type Output = QVariant;
+
+    /// Wrapper around [`at(int)`][method] method.
+    ///
+    /// [method]: https://doc.qt.io/qt-5/qlist.html#at
+    #[track_caller]
+    fn index(&self, key: QString) -> &QVariant {
+        unsafe {
+            cpp!([self as "QVariantMap*", key as "QString"] -> *const QVariant as "const QVariant*" {
+                auto x = self->constFind(key);
+                if (x == self->constEnd()) {
+                    return NULL;
+                } else {
+                    return &x.value();
+                }
+            }).as_ref()
+        }.expect("key not in the QVariant")
+    }
+}
+impl IndexMut<QString> for QVariantMap {
+    /// Wrapper around [`operator[](int)`][method] operator method.
+    ///
+    /// [method]: https://doc.qt.io/qt-5/qlist.html#operator-5b-5d
+    fn index_mut(&mut self, key: QString) -> &mut QVariant {
+        unsafe {
+            &mut *cpp!([self as "QVariantMap*", key as "QString"] -> *mut QVariant as "QVariant*" {
+                return &(*self)[key];
+            })
+        }
+    }
+}
+
+impl fmt::Debug for QVariantMap {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_map().entries(self.into_iter()).finish()
+    }
+}
+
+cpp_class!(unsafe struct QVariantMapIteratorInternal as "QVariantMap::iterator");
+
+/// Internal class used to iterate over a [`QVariantMap`][]
+///
+/// [`QVariantMap`]: ./struct.QVariantMap.html
+pub struct QVariantMapIterator<'a> {
+    map: &'a QVariantMap,
+    iterator: QVariantMapIteratorInternal,
+}
+
+impl<'a> QVariantMapIterator<'a> {
+    fn key(&self) -> Option<&'a QString> {
+        let iterator = &self.iterator;
+        unsafe {
+            cpp!([iterator as "QVariantMap::iterator*"] -> *mut QString as "const QString*" {
```

Comment:
```suggestion
            cpp!([iterator as "const QVariantMap::iterator*"] -> &QString as "const QString*" {
```
Not all the code respect that, but it would be better to respect the constness of things.
Like when we have a reference in rust, then use a const pointer in C++. and only use a non-const pointer in C++ if we have a &mut reference in rust.

Also you can just return rust references instead of pointers if you know it's not null or dangling.

Reply:
I cannot convert it to a reference since I am using `as_ref()` method to convert from pointer to `Option<&QString>`. I did convert it to *const QString though.

Reply:
ah yes, good point.  You could just use `-> Option<&QString> as "const QString*"`  because Option ensure that it use the niche.